modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/maven/AbstractMavenClassLoaderModelLoader.java

```diff
@@ -161,14 +166,42 @@ protected ClassLoaderModel createHeavyPackageClassLoaderModel(File artifactFile,
           .collect(toSet());
     }
 
-    loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    // TODO MULE-17114 retrieve this data from the json if present, if not then call this
+    populateNotImportingPackages(artifactFile, classLoaderModelBuilder, bundleDependencies);
+
     classLoaderModelBuilder.dependingOn(bundleDependencies);
 
     return classLoaderModelBuilder.build();
   }
 
+  protected void populateNotImportingPackages(File artifactFile, final ArtifactClassLoaderModelBuilder classLoaderModelBuilder,
+                                              Set<BundleDependency> bundleDependencies) {
+    final JarExplorer fileJarExplorer = buildJarExplorer();
+
+    // This is already filtering out mule-plugin dependencies,
+    // since for this case we explicitly need to consume the exported API from the plugin.
+    final List<URL> dependenciesArtifactsUrls = loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    for (URL dependencyArtifactUrl : dependenciesArtifactsUrls) {
+      final URI dependencyArtifactUri;
+      try {
+        dependencyArtifactUri = dependencyArtifactUrl.toURI();
+      } catch (URISyntaxException e) {
+        throw new MuleRuntimeException(e);
+      }
+
+      final JarInfo exploredJar = fileJarExplorer.explore(dependencyArtifactUri);
+
+      classLoaderModelBuilder.notImportingPackages(exploredJar.getPackages());
```

Comment:
This is not right to me. Packages contained on the artifact cannot be simply used without more analysis.
You need to calculate the artifact's API closure (ie: all the packages that must be exported from this artifact).
Packages in the closure that are already provided by the region cannot be overridden by the artifact. Otherwise you can get ClassCastExceptions.
You can only ignore packages from the region (the parent first ones) when they are not part of your exported API.

Besides that, there are other checks to do: no artifact can override packages from java and the SYSTEM_PACKAGES (ie: org.mule.runtime, com.mulesoft.mule.runtime)